### PR TITLE
RENO-1775: Handle late openings and early closings

### PR DESCRIPTION
**Pull request by nypl-wluisi**
[Jira Ticket](http://jira.nypl.org/browse/RENO-1775)

**This PR does the following:**
- Adds better timezone handling to `refineryApi` datasource.
- Renames util function `checkAlertsOpenStatus()` to `hasActiveClosing()`
- Adds test coverage for `hasActiveClosing()`
- Adds `naturalSort()` and related tests
- Adds `setTodaysHours()` util function and related tests

### Review Steps:
- [ ] Pull in branch
- [ ] Run tests `npm test`

### CODE REVIEW

File: src/utils/setTodaysHours.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    const endTimes = [todayHoursEnd];
+
+    // We have alerts, so map over them.
+    alerts.some(alert => {
+      // Check if closed_for key exists
+      if ('closed_for' in alert) {
+        // Check for start and end values
+        if (alert.applies.start && alert.applies.end) {
+          // Closing hours
+          const closingStartHours = dayjs(alert.applies.start).format('HH:mm');
```
Comment (nypl-wluisi) [inline src/utils/setTodaysHours.js:52]: Might need to include the tz method here: `.tz('America/New_York')` -- test locally by changing the timezone of your computer.

Comment (nypl-wluisi) [reply src/utils/setTodaysHours.js:52]: Try setting a global timezone for all dayjs

`dayjs.tz.setDefault("America/New_York")`

https://day.js.org/docs/en/timezone/set-default-timezone

File: src/utils/hasActiveClosing.test.js [hunk 2]
```diff
@@ -27,10 +27,10 @@ describe('checkAlertsOpenStatus', () => {
     }
   ];
 
-  test('One active temporary closing should return false.', () => {
+  test('One active temporary closing should return true.', () => {
     expect(
-      checkAlertsOpenStatus('2020-10-27T12:00:00-04:00', alertsTempClosingOne)
-    ).toBe(false);
+      hasActiveClosing('2020-10-27T12:00:00-04:00', alertsTempClosingOne, true)
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.test.js:32]: Remove third param

File: src/utils/hasActiveClosing.test.js [hunk 3]
```diff
@@ -76,21 +76,20 @@ describe('checkAlertsOpenStatus', () => {
     }
   ];
 
-  test('Many temporary closings, only 1 active, should return false.', () => {
+  test('Many temporary closings, only 1 active, should return true.', () => {
     expect(
-      checkAlertsOpenStatus('2020-10-27T12:00:00-04:00', alertsTempClosingMany)
-    ).toBe(false);
+      hasActiveClosing('2020-10-27T12:00:00-04:00', alertsTempClosingMany, true)
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.test.js:81]: remove third param

File: src/utils/hasActiveClosing.test.js [hunk 4]
```diff
@@ -109,10 +108,10 @@ describe('checkAlertsOpenStatus', () => {
     }
   ];
 
-  test('A single temporary closing with inactive hours should return true.', () => {
+  test('A single temporary closing with inactive hours should return false.', () => {
     expect(
-      checkAlertsOpenStatus('2020-10-27T12:00:00-04:00', alertsHoursMins)
-    ).toBe(true);
+      hasActiveClosing('2020-10-27T12:00:00-04:00', alertsHoursMins, true)
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.test.js:113]: remove third param

File: src/utils/hasActiveClosing.test.js [hunk 5]
```diff
@@ -141,9 +140,43 @@ describe('checkAlertsOpenStatus', () => {
     }
   ];
 
-  test('A single temporary closing with active hours should return false.', () => {
+  test('A single temporary closing with active hours should return true.', () => {
     expect(
-      checkAlertsOpenStatus('2020-10-27T12:05:00-04:00', alertsHoursMinsActive)
-    ).toBe(false);
+      hasActiveClosing('2020-10-27T12:05:00-04:00', alertsHoursMinsActive, true)
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.test.js:145]: remove third param

File: src/utils/naturalSort.test.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  const timesExpected = [
+    '01:00',
+    '02:00',
+    '10:00',
+    '12:00'
+  ];
+
+  test('Times are sorted alphanumerically', () => {
+    const timesNaturalSort = timesArray.sort(naturalSort);
+    console.log(timesNaturalSort);
```
Comment (jberube-nypl) [inline src/utils/naturalSort.test.js:20]: remove

File: src/utils/setTodaysHours.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+dayjs.extend(isBetween);
+// Utils
+import naturalSort from './naturalSort';
+
+/**
+ * Set today's hours using regular hours or modified hours using alerts closings.
+ *
+ * @param {object} now - current date object, unformatted.
+ * @param {object} regularHours - an object of regular hours data.
+ * @param {object} alerts - an object of alerts & closings data.
```
Comment (jberube-nypl) [inline src/utils/setTodaysHours.js:19]: @param {array} alerts - an array of alert objects containing alert info & closings data.

File: src/utils/setTodaysHours.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+  // Active closing, not extended closing
+  // Return hours as closed (null) or modify hours.
+  if (hasActiveClosing && !isExtendedClosing) {
+    // Add the regular hours values to sortable stacks of hours.
+    const startTimes = [todayHoursStart];
+    const endTimes = [todayHoursEnd];
+
+    // We have alerts, so map over them.
+    alerts.some(alert => {
```
Comment (jberube-nypl) [inline src/utils/setTodaysHours.js:52]: alerts.forEach(alert => {

**Review by jberube-nypl — CHANGES_REQUESTED** [review]

File: src/utils/filterByOpenNow.js [hunk 9]
```diff
@@ -15,7 +15,7 @@ function filterByOpenNow(now, locations) {
 
   return locations.reduce((accumlator, location) => {
```
[the remark this answers is not in the document]
Comment (nypl-wluisi) [reply src/utils/filterByOpenNow.js]: This is fixed.

File: src/utils/hasActiveClosing.js [hunk 10]
```diff
@@ -3,20 +3,20 @@ var isBetween = require('dayjs/plugin/isBetween');
...
+function hasActiveClosing(today, alerts) {
+  let activeClosing = false;
 
   // Check for any alerts.
   if (alerts === undefined || alerts.length === 0) {
     // No alerts, so set this status to true.
     // Other checks below will determine if the location is open.
-    alertsOpenStatus = true;
+    activeClosing = false;
   } else {
```
[the remark this answers is not in the document]
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.js]: This is fixed.

File: src/utils/filterByOpenNow.js [hunk 11]
```diff
@@ -15,7 +15,7 @@ function filterByOpenNow(now, locations) {
...
     // Alerts
-    const alertsOpenStatus = checkAlertsOpenStatus(today, location._embedded.alerts);
+    const isActiveClosing = hasActiveClosing(today, location._embedded.alerts, location.open);
```
[the remark this answers is not in the document]
Comment (nypl-wluisi) [reply src/utils/filterByOpenNow.js]: This is fixed.

File: src/utils/hasActiveClosing.test.js (hunk 2 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.test.js]: This is fixed.

File: src/utils/hasActiveClosing.test.js (hunk 3 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.test.js]: This is fixed.

File: src/utils/hasActiveClosing.test.js (hunk 4 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.test.js]: This is fixed.

File: src/utils/hasActiveClosing.test.js (hunk 5 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.test.js]: This is fixed.

File: src/utils/naturalSort.test.js (hunk 6 above)
Comment (nypl-wluisi) [reply src/utils/naturalSort.test.js]: This is fixed.

File: src/utils/setTodaysHours.js (hunk 7 above)
Comment (nypl-wluisi) [reply src/utils/setTodaysHours.js]: This is fixed.

File: src/utils/setTodaysHours.js (hunk 8 above)
Comment (nypl-wluisi) [reply src/utils/setTodaysHours.js]: This is fixed.

File: src/utils/hasActiveClosing.js [hunk 12]
```diff
@@ -3,37 +3,36 @@ var isBetween = require('dayjs/plugin/isBetween');
 dayjs.extend(isBetween);
 
 /**
- * Determines if a location is open by comparing today to alerts & closings data.
+ * Determines if a location has an active closing.
  *
  * @param {date} today - current datetime in ISO8601, i.e, 2020-10-27T12:00:00-04:00.
  * @param {object} alerts - an object of alerts & closings data.
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.js:9]: * @param {array} alerts - an array of alert objects containing alert info & closings data.

File: src/utils/hasActiveClosing.js [hunk 13]
```diff
@@ -3,37 +3,36 @@ var isBetween = require('dayjs/plugin/isBetween');
...
 
   // Check for any alerts.
-  if (alerts === undefined || alerts.length === 0) {
-    // No alerts, so set this status to true.
-    // Other checks below will determine if the location is open.
-    alertsOpenStatus = true;
-  } else {
+  if (alerts !== undefined && alerts.length > 0) {
     // We have alerts, so map over them.
     alerts.map(alert => {
```
Comment (jberube-nypl) [inline src/utils/hasActiveClosing.js:18]: alerts.forEach(alert => {

**Review by jberube-nypl — CHANGES_REQUESTED** [review]

File: src/utils/hasActiveClosing.js (hunk 12 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.js]: This is fixed.

File: src/utils/hasActiveClosing.js (hunk 13 above)
Comment (nypl-wluisi) [reply src/utils/hasActiveClosing.js]: This is fixed.

**Review by jberube-nypl — APPROVED** [review]

**Pull request merged**